### DISCUSSION TITLE
Newbie question on using ImageSharp

### DISCUSSION THREAD

**Discussion by jkowalski0058** [SixLabors/ImageSharp#1807] [Q&A] [answered]
I hope I'm in the right place to ask newbie questions (my first post), please redirect me to the right place if not.
I have legacy code doing 2D graphics, C# VS2019 .NET Framework 4.5 that I am porting over to .NET 5.0 C# VS2019.
I just got ImageSharp and I can open a png file and display it in WPF using ImageSharp, Yay! :-)
Below are some first examples to tackle and my web searches so far have failed me.
How do I convert the following to C# code using ImageSharp?

//============================================================
        private static readonly ColorPalette GrayLut;
//============================================================

            // temporary bitmap
            var bmp = new Bitmap( 64, 64, System.Drawing.Imaging.PixelFormat.Format8bppIndexed );

            GrayLut = bmp.Palette;

            for ( int i = 0; i < 256; i++ )
            {
                GrayLut.Entries[i] = System.Drawing.Color.FromArgb( 255, i, i, i );
            }
//=============================================================
        public static Bitmap Create8BitBitmap( byte[] imageDataBytes, int width, int height )
        {
            System.Drawing.Imaging.PixelFormat pixelFormat = System.Drawing.Imaging.PixelFormat.Format8bppIndexed;

            var b = new Bitmap( width, height, pixelFormat );

            b.Palette = GrayLut;

            var boundsRect = new Rectangle( 0, 0, width, height );

            BitmapData bmpData = b.LockBits( boundsRect, ImageLockMode.WriteOnly, pixelFormat );

            Marshal.Copy( imageDataBytes, 0, (IntPtr) bmpData.Scan0, imageDataBytes.Length );

            b.UnlockBits(bmpData);

            return b;
        }

Thanks in advance for any help

**Comment by tocsoft** [comment, marked as answer]
I think you want from the code above to create a grey scale image where you have a source `byte[]` made up of pixel data where each byte represents a single grey scale pixel.

I believe this is just the `L8` pixel format (no need for palette stuff)

So I think this can all end up as a one liner of 
``` c#
var greyImg = Image.LoadPixelData<L8>(imageDataBytes, width, height); // don't forget to dispose of this when your finished.
```

if you plan to then post process the image more and need a full color space you can then use 
```c#
var rgbImage = greyImg.CloneAs<Rgba32>(); // now both greyImage and rgbImage are completely separate and both need disposing
````
which will create a new image expanded to the full `Rgba32` color space

**Reply by jkowalski0058** [reply]
In our legacy app all of our images are from a monochrome camera and we need the byte array of pixels to search for objects. How do I open a PNG file and convert that into a byte array?
`            Image image = Image.Load<L8>(imageFilename);

            Size size  = image.Size();
            int nBytes = size.width * size.height;

            // create container for pixel byte array
            byte[] imageDataBytes = new byte[nBytes];
==>   How do I copy in all pixel bytes?
`

**Reply by jkowalski0058** [reply]
I tried this from [an example](https://docs.sixlabors.com/articles/imagesharp/pixelbuffers.html?q=TryGetSinglePixelSpan) but it complains about TryGetSinglePixelSpan

            Image image = Image.Load<L8>(imageFilename);
            Size size  = image.Size();
            int width  = size.Width;
            int height = size.Height;
            int nBytes = width * height;
            byte[] imageDataBytes = new byte[nBytes];
            if (image.TryGetSinglePixelSpan(out var pixelSpan))
            {
                imageDataBytes = MemoryMarshal.AsBytes(pixelSpan()).ToArray();
            }

Error CS1061 'Image' does not contain a definition for 'TryGetSinglePixelSpan' and no accessible extension method 'TryGetSinglePixelSpan' accepting a first argument of type 'Image' could be found (are you missing a using directive or an assembly reference?)

**Reply by jkowalski0058** [reply]
I've been searching the web for examples but not having luck. is there a PDF or a book I can buy with examples on how to use the library?

**Reply by tocsoft** [reply]
The library is setup to be used in one of 2 higher level modes, basic image manipulation (using the processors in a pixel format agnostic fashion), and lower level pixel access (where you need to opt into a specific pixel format). This switch in modes is based on the type of `Image` class your code uses.

`Image` is for the higher level processing where you are leaving the internals of pixels up to us.
`Image<TPixel>` (which inherits from `Image`) for when you want low level access to the individual pixel data, via pixel spans and indexers etc. (and `TryGetSinglePixelSpan()` along with them.)

So inconclusion The `TryGetSinglePixelSpan` only exists on the `Image<TPixel>` type not `Image`. 

In your code above you are correctly loading the image into an `Image<L8>` but immediately downcast it to `Image` which hid the pixel level APIs from you. 

If you just change 
```c#
Image image = Image.Load<L8>(imageFilename);
```
to 
```c#
Image<L8> image = Image.Load<L8>(imageFilename); // note the extra <L8> in the variable declaration
```
then the method should show up.

We don't have any form of PDF/book for the library our [docs site](https://docs.sixlabors.com/) is currently the best we've got

**Reply by jkowalski0058** [reply]
Thank you, I'm getting there.

**Comment by antonfirsov** [comment]
@jkowalski0058 since `TryGetSinglePixelSpan` is about to go in ImageSharp 2.0 (#1739), I would like to understand your use case better in order to design the best replacement API/s, so I have a few questions:

Why do you need to copy the pixels to a new array? How is the array being consume? Is it passed to some other library? (If yes: in form of `byte[]` or a pointer?) Can't you do one of the following instead?
- Pass around `Span<L8>` or `Span<byte>` 
- Pass around the `Image<L8>` object, and process it row-by-row, when needed

**Reply by jkowalski0058** [reply]
Here is my example code. I load the image into Image<L8> OK and it correctly identifies the size as 2240 x 2440 which should provide a byte array of 5017600 which my legacy app does and is able to process. However, after different attempts with ImageSharp to get a byte array I never get the correct length and of course crashes in my function CalculateImageQuality() due to index out of bounds.
![ImageSharpAttemptAtByteArray](https://user-images.githubusercontent.com/77497339/140436792-24438ad0-c089-4b86-a2e8-edc50e5fcc0f.png)


Outputs:
Width = 2240, Height = 2240, Number of bytes = 5017600
Byte array length = 2086628

Commenting out these three lines which I do not understand but got 
from an example somewhere
// pngEncoder.CompressionLevel = PngCompressionLevel.Level5;
// pngEncoder.BitDepth = PngBitDepth.Bit8;
// pngEncoder.ColorType = PngColorType.Grayscale;
Outputs:
Width = 2240, Height = 2240, Number of bytes = 5017600
Byte array length = 1357690

**Reply by tocsoft** [reply]
your saving as PNG, a PNG is way more that the raw pixel data, its also a compressed format thus being smaller... why did you changed your code from earlier when you correctly extracted the pixel data using?
```c#
byte[] imageDataBytes = new byte[nBytes];
if (image.TryGetSinglePixelSpan(out var pixelSpan))
{
    imageDataBytes = MemoryMarshal.AsBytes(pixelSpan()).ToArray();
}
```

**Comment by jkowalski0058** [comment]
I truly appreciate the support. I have legacy code currently maintained and built with VS2019 version 16.11.5, C# targeting .NET Framework 4.5. The images are captured by a monochrome camera C4410 20 MP CMOS Camera (imperx.com)<https://www.imperx.com/cmos-cameras/C4410/>. We take one image and search through the pixels locating objects that look like our target pattern. We take a second image to illuminate the fluorescence tags and then for each object location we evaluate the relative intensity to determine if the target is activated with fluorescence or not.

I have created a new project with .NET 5.0 VS2019 C# which I understand no longer supports the Microsoft imaging tools so I can't compile to load images to process the pixels as bytes.  In my limited search I found your library which seems popular and am trying to get it to work and evaluate. I've attached one legacy file which does most of the image file loading and converting to byte arrays. Once I have a byte array in the same format as my legacy app, I can pass the byte array to lower-level functions that will process through the pixels, locate objects, and evaluate intensities.

My graphics expertise is very limited and am on a steep learning curve, appreciate any help.

**Comment by jkowalski0058** [comment]
Mostly because I have no idea what I’m doing ☹. I don’t understand how the legacy code works with graphics files so I’m really struggling to figure that out too. When I paste in your code into my new .NET 5 project I get compiler error (see attached image).

**Reply by jkowalski0058** [reply]
![GetBytes](https://user-images.githubusercontent.com/77497339/140481275-2981caeb-3173-42b5-844c-5c540e18c731.png)

**Comment by tocsoft** [comment]
sorry my bad, I assumed the previous code example I copied from was 100% correct rather than have a typo in it. If you use intellisense to look at what type `pixelSpan` is you will see its a `Span<L8>` and thus isn't a method.
```c#
imageDataBytes = MemoryMarshal.AsBytes(pixelSpan()).ToArray();
```
should actually be
```c#
imageDataBytes = MemoryMarshal.AsBytes(pixelSpan).ToArray(); // no brackets after pixelSpan its a type not 
```

**Reply by jkowalski0058** [reply]
You are the Master! It works and I was able to feed the byte array into my legacy function CalculateImageQuality() which returned the exact same values. Very excited to move on now.
-Grasshopper
![GetRawByesForLegacyFunctions](https://user-images.githubusercontent.com/77497339/140487346-8c1e6cc4-fd96-4133-a1bc-4e4b9d1c17e7.png)

**Reply by jkowalski0058** [reply]
Just FYI I had tried that Memory Marshall technique the other day as I saw it in an example but when it didn't compile I wasn't smart enough to fix it and continued my search for the magic key :-). Thank you so much for the help.

**Reply by antonfirsov** [reply]
@jkowalski0058 you should always `Dispose()` the image if you no longer use it! Put your top-level `Image<L8>` consumer code into a using statement, or just put a `using` keyword before `Image<L8> image = Image.Load<L8>(imageFilename);` if you can use C# 8: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using#using-declaration
